### PR TITLE
RHINENG-12670: Add script for publishing hosts

### PR DESCRIPTION
**Pull request by thearifismail**
rh-pre-commit.version: 2.2.0
rh-pre-commit.check-secrets: ENABLED

# Overview

This PR is being created to address [RHINENG-12670](https://issues.redhat.com/browse/RHINENG-12670).
It checks the host-inventory database for a publication named `hbi_hosts_pub`.  If the publication is not found, it creates it.

## PR Checklist

- [ ] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Include raw query examples in the PR description, if adding/modifying SQL query
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: inv_publish_hosts.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+def _excepthook(logger, type, value, traceback):
+    logger.exception("Inventory migration failed", exc_info=value)
+
+
+def run(logger, session, application):
+    with application.app.app_context():
+        logger.info(f"Checking for publication using the following SQL statement:\n\t{CHECK_PUBLICATION}")
+        result = session.execute(sa_text(CHECK_PUBLICATION))
+        found = result.cursor.fetchone()[0]
```
Comment (chambridge) [inline inv_publish_hosts.py:46]: Does this work if the publication doesn't exist? Feels like you would be trying to get the zeroth index of `None` in that case.

File: deploy/clowdapp.yml [hunk 2]
```diff
@@ -932,6 +932,79 @@ objects:
...
+          - name: UNLEASH_URL
+            value: ${UNLEASH_URL}
+          - name: UNLEASH_TOKEN
+            valueFrom:
+              secretKeyRef:
+                name: ${UNLEASH_SECRET_NAME}
+                key: CLIENT_ACCESS_TOKEN
+                optional: true
+          - name: BYPASS_UNLEASH
+            value: ${BYPASS_UNLEASH}
```
Comment (chambridge) [inline deploy/clowdapp.yml:999]: I don't think this job needs all of these environment values for what its doing. We should just remove the unnecessary items.

**Review by chambridge — CHANGES_REQUESTED** [review]
Some suggestions and comments.

File: inv_publish_hosts.py (hunk 1 above)
Comment (thearifismail) [reply inv_publish_hosts.py:46]: When the index does not exist, the `fetchone()[0]` returns ` `False`

**Review by thearifismail — COMMENTED** [review]
@chambridge addressed your comments.

File: inv_publish_hosts.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        if found:
+            logger.info(f'Publication "{PUBLICATION_NAME}" found!')
```
Comment (chambridge) [inline inv_publish_hosts.py:48]: Can we add a check here to make sure all the replication slots are active?
`select slot_name, active from pg_replication_slots`
active column will be a true or false boolean. If the value is false can we print out the slot_name that is not active this could highlight a replication failure that needs to be resolved.

After printing out any inactive replication slots, if one exists than can we exit the Job with a failure `exit(1)` that way the Job alert will bring our attention the replication issue if it exists.

**Review by chambridge — CHANGES_REQUESTED** [review]
A few more changes needed here.

Also this doesn't currently run when deployed as you need to add `inv_publish_hosts.py` to the Dockerfile.

The job is failing in ephemeral:
```
./run_command.sh: line 12: ./inv_publish_hosts.py: No such file or directory
```
https://ci.ext.devshift.net/job/RedHatInsights-insights-host-inventory-pr-check/5521/artifact/artifacts/k8s_artifacts/ephemeral-czeej6/logs/host-inventory-syndicator-28890400-7qpgv_host-inventory-syndicator.log

File: inv_publish_hosts.py (hunk 3 above)
Comment (thearifismail) [reply inv_publish_hosts.py:49]: Done!

**Review by thearifismail — COMMENTED** [review]
@chambridge addressed your comments.

File: Dockerfile [hunk 4]
```diff
@@ -28,6 +28,7 @@ COPY gunicorn.conf.py gunicorn.conf.py
 COPY host_reaper.py host_reaper.py
 COPY host_synchronizer.py host_synchronizer.py
 COPY inv_mq_service.py inv_mq_service.py
+COPY inv_publish_hosts.py inv_publish_hosts
```
Comment (chambridge) [inline Dockerfile:31]:
```suggestion
COPY inv_publish_hosts.py inv_publish_hosts.py
```
Minor typo here. You can see the job is still currently failing in the tests.

**Review by chambridge — APPROVED** [review]
lgtm

**Review by strider — APPROVED** [review]
LGTM
